File: archive/cdktf-java/Pr3239/lib/src/main/java/app/constructs/ApiGatewayConstruct.java
package app.constructs;

import com.hashicorp.cdktf.providers.aws.api_gateway_account.ApiGatewayAccount;
import com.hashicorp.cdktf.providers.aws.api_gateway_account.ApiGatewayAccountConfig;
import com.hashicorp.cdktf.providers.aws.api_gateway_deployment.ApiGatewayDeployment;
import com.hashicorp.cdktf.providers.aws.api_gateway_deployment.ApiGatewayDeploymentConfig;
import com.hashicorp.cdktf.providers.aws.api_gateway_integration.ApiGatewayIntegration;
import com.hashicorp.cdktf.providers.aws.api_gateway_integration.ApiGatewayIntegrationConfig;
import com.hashicorp.cdktf.providers.aws.api_gateway_method.ApiGatewayMethod;
import com.hashicorp.cdktf.providers.aws.api_gateway_method.ApiGatewayMethodConfig;
import com.hashicorp.cdktf.providers.aws.api_gateway_method_settings.ApiGatewayMethodSettings;
import com.hashicorp.cdktf.providers.aws.api_gateway_method_settings.ApiGatewayMethodSettingsConfig;
import com.hashicorp.cdktf.providers.aws.api_gateway_method_settings.ApiGatewayMethodSettingsSettings;
import com.hashicorp.cdktf.providers.aws.api_gateway_resource.ApiGatewayResource;
import com.hashicorp.cdktf.providers.aws.api_gateway_resource.ApiGatewayResourceConfig;
import com.hashicorp.cdktf.providers.aws.api_gateway_rest_api.ApiGatewayRestApi;
import com.hashicorp.cdktf.providers.aws.api_gateway_rest_api.ApiGatewayRestApiConfig;
import com.hashicorp.cdktf.providers.aws.api_gateway_rest_api.ApiGatewayRestApiEndpointConfiguration;
import com.hashicorp.cdktf.providers.aws.api_gateway_stage.ApiGatewayStage;
import com.hashicorp.cdktf.providers.aws.api_gateway_stage.ApiGatewayStageConfig;
import com.hashicorp.cdktf.providers.aws.iam_role.IamRole;
import com.hashicorp.cdktf.providers.aws.iam_role.IamRoleConfig;
import com.hashicorp.cdktf.providers.aws.iam_role_policy_attachment.IamRolePolicyAttachment;
import com.hashicorp.cdktf.providers.aws.iam_role_policy_attachment.IamRolePolicyAttachmentConfig;
import com.hashicorp.cdktf.providers.aws.lambda_permission.LambdaPermission;
import com.hashicorp.cdktf.providers.aws.lambda_permission.LambdaPermissionConfig;
import software.constructs.Construct;

import java.util.Arrays;
import java.util.List;

public class ApiGatewayConstruct extends BaseConstruct {
    private final ApiGatewayRestApi api;

    public ApiGatewayConstruct(final Construct scope, final String id, final String lambdaArn, final String lambdaInvokeArn) {
        super(scope, id);

        // Create CloudWatch Logs role for API Gateway
        IamRole apiGatewayCloudWatchRole = new IamRole(this, "api-cloudwatch-role", IamRoleConfig.builder()
                .name(resourceName("APIGatewayCloudWatchLogsRole"))
                .assumeRolePolicy("""
                        {
                          "Version": "2012-10-17",
                          "Statement": [
                            {
                              "Effect": "Allow",
                              "Principal": {
                                "Service": "apigateway.amazonaws.com"
                              },
                              "Action": "sts:AssumeRole"
                            }
                          ]
                        }
                        """)
                .tags(getTagsWithName("APICloudWatchRole"))
                .build());

        // Attach CloudWatch Logs policy to API Gateway role
        new IamRolePolicyAttachment(this, "api-cloudwatch-policy", IamRolePolicyAttachmentConfig.builder()
                .role(apiGatewayCloudWatchRole.getName())
                .policyArn("arn:aws:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs")
                .build());

        // Configure API Gateway account settings for CloudWatch Logs
        ApiGatewayAccount apiGatewayAccount = new ApiGatewayAccount(this, "api-account", ApiGatewayAccountConfig.builder()
                .cloudwatchRoleArn(apiGatewayCloudWatchRole.getArn())
                .build());

        // Create REST API
        this.api = new ApiGatewayRestApi(this, "rest-api", ApiGatewayRestApiConfig.builder()
                .name(resourceName("API"))
                .description("Serverless Demo API Gateway")
                .endpointConfiguration(ApiGatewayRestApiEndpointConfiguration.builder()
                        .types(List.of("REGIONAL"))
                        .build())
                .tags(getTagsWithName("API"))
                .build());

        // Create resource
        ApiGatewayResource resource = new ApiGatewayResource(this, "api-resource",
                ApiGatewayResourceConfig.builder()
                        .restApiId(api.getId())
                        .parentId(api.getRootResourceId())
                        .pathPart("items")
                        .build());

        // Create GET method
        ApiGatewayMethod getMethod = new ApiGatewayMethod(this, "get-method",
                ApiGatewayMethodConfig.builder()
                        .restApiId(api.getId())
                        .resourceId(resource.getId())
                        .httpMethod("GET")
                        .authorization("NONE")
                        .build());

        // Create POST method
        ApiGatewayMethod postMethod = new ApiGatewayMethod(this, "post-method",
                ApiGatewayMethodConfig.builder()
                        .restApiId(api.getId())
                        .resourceId(resource.getId())
                        .httpMethod("POST")
                        .authorization("NONE")
                        .build());

        // Create Lambda integrations
        ApiGatewayIntegration getIntegration = new ApiGatewayIntegration(this, "get-integration", ApiGatewayIntegrationConfig.builder()
                .restApiId(api.getId())
                .resourceId(resource.getId())
                .httpMethod(getMethod.getHttpMethod())
                .integrationHttpMethod("POST")
                .type("AWS_PROXY")
                .uri(lambdaInvokeArn)
                .build());

        ApiGatewayIntegration postIntegration = new ApiGatewayIntegration(this, "post-integration", ApiGatewayIntegrationConfig.builder()
                .restApiId(api.getId())
                .resourceId(resource.getId())
                .httpMethod(postMethod.getHttpMethod())
                .integrationHttpMethod("POST")
                .type("AWS_PROXY")
                .uri(lambdaInvokeArn)
                .build());

        // Create deployment
        ApiGatewayDeployment deployment = new ApiGatewayDeployment(this, "api-deployment",
                ApiGatewayDeploymentConfig.builder()
                        .restApiId(api.getId())
                        .dependsOn(Arrays.asList(getMethod, postMethod, getIntegration, postIntegration))
                        .build());

        // Create stage with logging
        ApiGatewayStage stage = new ApiGatewayStage(this, "api-stage", ApiGatewayStageConfig.builder()
                .deploymentId(deployment.getId())
                .restApiId(api.getId())
                .stageName("prod")
                .xrayTracingEnabled(true)
                .tags(getTagsWithName("APIStage"))
                .build());

        // Configure method settings for logging (depends on account configuration)
        new ApiGatewayMethodSettings(this, "method-settings", ApiGatewayMethodSettingsConfig.builder()
                .restApiId(api.getId())
                .stageName(stage.getStageName())
                .methodPath("*/*")
                .settings(ApiGatewayMethodSettingsSettings.builder()
                        .loggingLevel("INFO")
                        .dataTraceEnabled(true)
                        .metricsEnabled(true)
                        .build())
                .dependsOn(List.of(apiGatewayAccount))
                .build());

        // Grant API Gateway permission to invoke Lambda
        new LambdaPermission(this, "api-lambda-permission", LambdaPermissionConfig.builder()
                .statementId("AllowAPIGatewayInvoke")
                .action("lambda:InvokeFunction")
                .functionName(lambdaArn)
                .principal("apigateway.amazonaws.com")
                .sourceArn(api.getExecutionArn() + "/*/*")
                .build());
    }

    public String getApiName() {
        return api.getName();
    }

    public String getApiId() {
        return api.getId();
    }
}

File: archive/cdktf-py/Pr6581/lib/api_gateway_construct.py
"""API Gateway REST API."""
from cdktf_cdktf_provider_aws.api_gateway_deployment import \
    ApiGatewayDeployment
from cdktf_cdktf_provider_aws.api_gateway_integration import \
    ApiGatewayIntegration
from cdktf_cdktf_provider_aws.api_gateway_method import ApiGatewayMethod
from cdktf_cdktf_provider_aws.api_gateway_resource import ApiGatewayResource
from cdktf_cdktf_provider_aws.api_gateway_rest_api import ApiGatewayRestApi
from cdktf_cdktf_provider_aws.api_gateway_stage import ApiGatewayStage
from cdktf_cdktf_provider_aws.data_aws_region import DataAwsRegion
from cdktf_cdktf_provider_aws.lambda_permission import LambdaPermission
from constructs import Construct


class ApiGatewayConstruct(Construct):
    """API Gateway for payment validation."""

    def __init__(self, scope: Construct, id: str, environment_suffix: str, validator_function):
        super().__init__(scope, id)

        # Get current region
        current_region = DataAwsRegion(self, "current-region")

        # REST API
        api = ApiGatewayRestApi(
            self, "api",
            name=f"payment-api-{environment_suffix}-ef",
            description="Payment processing API",
            tags={
                "Name": f"payment-api-{environment_suffix}-ef",
                "Environment": environment_suffix,
                "Project": "payment-processing",
                "CostCenter": "engineering"
            }
        )

        # /payments resource
        payments_resource = ApiGatewayResource(
            self, "payments-resource",
            rest_api_id=api.id,
            parent_id=api.root_resource_id,
            path_part="payments"
        )

        # POST method
        method = ApiGatewayMethod(
            self, "post-method",
            rest_api_id=api.id,
            resource_id=payments_resource.id,
            http_method="POST",
            authorization="NONE"
        )

        # Lambda integration
        integration = ApiGatewayIntegration(
            self, "lambda-integration",
            rest_api_id=api.id,
            resource_id=payments_resource.id,
            http_method=method.http_method,
            integration_http_method="POST",
            type="AWS_PROXY",
            uri=validator_function.invoke_arn
        )

        # Lambda permission
        LambdaPermission(
            self, "api-lambda-permission",
            statement_id="AllowAPIGatewayInvoke",
            action="lambda:InvokeFunction",
            function_name=validator_function.function_name,
            principal="apigateway.amazonaws.com",
            source_arn=f"{api.execution_arn}/*/*"
        )

        # Deployment
        deployment = ApiGatewayDeployment(
            self, "deployment",
            rest_api_id=api.id,
            depends_on=[integration]
        )

        # Stage
        stage = ApiGatewayStage(
            self, "stage",
            deployment_id=deployment.id,
            rest_api_id=api.id,
            stage_name="prod",
            tags={
                "Name": f"payment-api-prod-{environment_suffix}-ef",
                "Environment": environment_suffix
            }
        )

        self.api_url = f"https://{api.id}.execute-api.{current_region.id}.amazonaws.com/{stage.stage_name}/payments"
